pkg/controllers/generic/syncer.go

```diff
@@ -0,0 +1,310 @@
+}
+
+// TranslateMetadata converts the virtual object into a physical object
+func (f *exporter) TranslateMetadata(vObj client.Object) client.Object {
+	pObj := f.NamespacedTranslator.TranslateMetadata(vObj)
+	labels := pObj.GetLabels()
+	if labels == nil {
+		labels = map[string]string{}
+	}
+	labels[controlledByLabel] = f.getControllerID()
+	pObj.SetLabels(labels)
+	return pObj
+}
+
+func (f *exporter) IsManaged(pObj client.Object) (bool, error) {
+	if !translate.IsManaged(pObj) {
+		return false, nil
+	}
+
+	return !f.isExcluded(pObj), nil
+}
+
+func isControlled(obj client.Object) bool {
+	return obj.GetLabels() != nil && obj.GetLabels()[controlledByLabel] != ""
+}
+
+func (f *exporter) objectMatches(obj client.Object) bool {
+	return f.selector == nil || !f.selector.Matches(labels.Set(obj.GetLabels()))
+}
+
+type virtualToHostNameResolver struct {
+	namespace       string
+	targetNamespace string
+}
+
+func (r *virtualToHostNameResolver) TranslateName(name string, regex *regexp.Regexp, _ string) (string, error) {
+	return r.TranslateNameWithNamespace(name, r.namespace, regex, "")
+}
+
+func (r *virtualToHostNameResolver) TranslateNameWithNamespace(name string, namespace string, regex *regexp.Regexp, _ string) (string, error) {
+	if regex != nil {
+		return patchesregex.ProcessRegex(regex, name, func(name, ns string) types.NamespacedName {
+			// if the regex match doesn't contain namespace - use the namespace set in this resolver
+			if ns == "" {
+				ns = namespace
+			}
+
+			translate.Default.PhysicalNamespace(namespace)
+			return types.NamespacedName{Namespace: r.targetNamespace, Name: translate.Default.PhysicalName(name, ns)}
+		}), nil
+	} else {
+		return translate.Default.PhysicalName(name, namespace), nil
+	}
+}
+
+func (r *virtualToHostNameResolver) TranslateLabelExpressionsSelector(selector *metav1.LabelSelector) (*metav1.LabelSelector, error) {
+	var s *metav1.LabelSelector
+	if selector != nil {
+		s = &metav1.LabelSelector{MatchLabels: map[string]string{}}
+		for k, v := range selector.MatchLabels {
+			s.MatchLabels[k] = v
+		}
+		if len(selector.MatchExpressions) > 0 {
+			s.MatchExpressions = []metav1.LabelSelectorRequirement{}
+			for i, r := range selector.MatchExpressions {
+				s.MatchExpressions[i] = metav1.LabelSelectorRequirement{
+					Key:      r.Key,
+					Operator: r.Operator,
+					Values:   r.Values,
+				}
+			}
+		}
+		s.MatchLabels[translate.NamespaceLabel] = r.namespace
+		s.MatchLabels[translate.MarkerLabel] = translate.Suffix
+	}
+	return s, nil
+}
+
+func (r *virtualToHostNameResolver) TranslateLabelKey(key string) (string, error) {
+	return key, nil
+}
+
+func (r *virtualToHostNameResolver) TranslateLabelSelector(selector map[string]string) (map[string]string, error) {
+	s := map[string]string{}
+	if selector != nil {
+		for k, v := range selector {
+			s[k] = v
+		}
+		s[translate.NamespaceLabel] = r.namespace
+		s[translate.MarkerLabel] = translate.Suffix
+	}
+	return s, nil
```

Comment:
This should probably use `translate.Default.TranslateLabelSelector` - the argument types don't match, so some modifications are in order of course.